components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/impl/CacheBackedOAuthConsumerAppDAO.java

```diff
@@ -0,0 +1,202 @@
+                if (log.isDebugEnabled()) {
+                    log.debug("OAuth application for consumer key: " + consumerKey + " has been added to the " +
+                            "OAuthConsumerAppCache.");
+                }
+            }
+        }
+
+        return oAuthAppDO;
+    }
+
+    @Override
+    public OAuthAppDO getAppInformationByAppName(String appName) throws OAuthConsumerAppException {
+
+        OAuthAppDO oAuthAppDO = oAuthConsumerAppDAO.getAppInformationByAppName(appName);
+        // since we have persisted the OAuth2 app data against the appname, we are storing it in the cache against
+        // the consumer key for future references
+        if (OAuthConsumerAppCache.getInstance()
+                .getValueFromCache(oAuthAppDO.getOauthConsumerKey()) == null) {
+            OAuthConsumerAppCache.getInstance().addToCache(oAuthAppDO.getOauthConsumerKey(), oAuthAppDO);
+            if (log.isDebugEnabled()) {
+                log.debug("OAuth application for application: " + appName + " has been added to the " +
+                        "OAuthConsumerAppCache wiht the consumer key: " + oAuthAppDO.getOauthConsumerKey());
+            }
+        }
+        return oAuthAppDO;
+    }
+
+    @Override
+    public OAuthAppDO[] getOAuthConsumerAppsOfUser(String username, int tenantId) throws OAuthConsumerAppException {
+
+        return oAuthConsumerAppDAO.getOAuthConsumerAppsOfUser(username, tenantId);
+    }
+
+    @Override
+    public String getOAuthConsumerSecret(String consumerKey) throws OAuthConsumerAppException {
+
+        OAuthAppDO oAuthAppDO = OAuthConsumerAppCache.getInstance().getValueFromCache(consumerKey);
+        if (oAuthAppDO != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("OAuthConsumerAppCache exists for the OAuth application with consumer key: " +
+                        consumerKey);
+            }
+            return oAuthAppDO.getOauthConsumerSecret();
+        }
+        return oAuthConsumerAppDAO.getOAuthConsumerSecret(consumerKey);
+    }
+
+    @Override
+    public String getConsumerApplicationOwnerName(String consumerKey) throws OAuthConsumerAppException {
+
+        OAuthAppDO oAuthAppDO = OAuthConsumerAppCache.getInstance().getValueFromCache(consumerKey);
+        if (oAuthAppDO != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("OAuthConsumerAppCache exists for the OAuth application with consumer key: " +
+                        consumerKey);
+            }
+            return oAuthAppDO.getAppOwner().getUserName();
+        }
+        return oAuthConsumerAppDAO.getConsumerApplicationOwnerName(consumerKey);
+    }
+
+    @Override
+    public void updateOAuthConsumerApplication(OAuthAppDO oauthAppDO) throws OAuthConsumerAppException {
+
+        oAuthConsumerAppDAO.updateOAuthConsumerApplication(oauthAppDO);
+        OAuthConsumerAppCache.getInstance().addToCache(oauthAppDO.getOauthConsumerKey(), oauthAppDO);
```

Comment:
Have we checked whether adding to the cache will send cache invalidation msg in clustering?